libc/strtod.c

```diff
@@ -0,0 +1,189 @@
+	TEST_ASSERT_EQUAL_DOUBLE(1.00000000000000022204, strtod("1.00000000000000022204", NULL));
+	TEST_ASSERT_EQUAL_DOUBLE(2.225073858507201E-308, strtod("2.225073858507201E-308", NULL));
+}
+
+
+TEST(strtod_family, remaining_string)
+{
+	char *tmp = NULL;
+
+	char test1[] = "    1.797693134862315e+308";
+	strtod(test1, &tmp);
+	TEST_ASSERT_EQUAL_STRING("\0", tmp);
+
+	tmp = NULL;
+	char test2[] = "3.12345Alma mam lkorta";
+	strtod(test2, &tmp);
+	TEST_ASSERT_EQUAL_STRING("Alma mam lkorta", tmp);
+
+	tmp = NULL;
+	char test3[] = "    14999   ";
+	strtod(test3, &tmp);
+	TEST_ASSERT_EQUAL_STRING("   ", tmp);
+
+	tmp = NULL;
+	char test4[] = " .875Ealao";
+	strtod(test4, &tmp);
+	TEST_ASSERT_EQUAL_STRING("Ealao", tmp);
+
+	tmp = NULL;
+	char test5[] = "aaaaaaaaaa";
+	strtod(test5, &tmp);
+	TEST_ASSERT_EQUAL_STRING(test5, tmp);
+
+
+	tmp = NULL;
+	char test6[] = "-aaaaaaaaaa";
+	strtod(test6, &tmp);
+	TEST_ASSERT_EQUAL_STRING(test6, tmp);
+
+	tmp = NULL;
+	char test7[] = "+aaaaaaaaaa";
+	strtod(test7, &tmp);
+	TEST_ASSERT_EQUAL_STRING(test7, tmp);
+
+	tmp = NULL;
+	char test8[] = "";
+	strtod(test8, &tmp);
+	TEST_ASSERT_EQUAL_STRING(test8, tmp);
+}
+
+TEST(strtod_family, too_long_numbers)
+{
+	const char num1[] =
+		"17976931348623157081452742373170435679807056752584499659891747680315"
+		"72607800285387605895586327668781715404589535143824642343213268894641"
+		"82768467546703537516986049910576551282076245490090389328944075868508"
+		"45513394230458323690322294816580855933212334827479782620414472316873"
+		"8177180919299881250404026184124858368123111";
+	TEST_ASSERT_DOUBLE_IS_INF(strtod(num1, NULL));
+
+	const char num2[] =
+		"       17976931348623157081452742373170435679807056752584499659891747680315"
+		"72607800285387605895586327668781715404589535143824642343213268894641"
+		"82768467546703537516986049910576551282076245490090389328944075868508"
+		"45513394230458323690322294816580855933212334827479782620414472316873"
+		"817718091929988125040402618412485836812311122222";
+	TEST_ASSERT_DOUBLE_IS_INF(strtod(num2, NULL));
+
+	const char num3[] =
+		"27976931348623157081452742373170435679807056752584499659891747680315"
+		"72607800285387605895586327668781715404589535143824642343213268894641"
+		"82768467546703537516986049910576551282076245490090389328944075868508"
+		"45513394230458323690322294816580855933212334827479782620414472316873"
+		"8177180919299881250404026184124858368";
+	TEST_ASSERT_DOUBLE_IS_INF(strtod(num3, NULL));
+}
+
+TEST(strtod_family, errno_test)
+{
+	errno = 0;
+	const char num1[] =
+		"17976931348623157081452742373170435679807056752584499659891747680315"
+		"72607800285387605895586327668781715404589535143824642343213268894641"
+		"82768467546703537516986049910576551282076245490090389328944075868508"
+		"45513394230458323690322294816580855933212334827479782620414472316873"
+		"8177180919299881250404026184124858368123111";
+	strtod(num1, NULL);
```

Comment:
Hmm, shouldn't we also assert that `strtod` retval is `NAN`?

Reply:
It is very reasonable to check retval, i will add it. However, i think retval should be equal to `\0` character.